### PR TITLE
Simplified setup.py A LOT

### PR DESCRIPTION
**Pull request by nils-werner**
Warning: This is probably a scary aggressive pull request. And I might have some peculiar ideas about updates and assuming an up-to-date target system.

I've found many antipatterns in your `setup.py` file:

 - Differentiating between distutils/setuptools (including versions of them)
 - Managing dependencies manually (instead of letting PIP do the heavy lifting)
 - Doing some weird git tag management
 - Manually cythonizing code (instead of letting Cython do the job)
 - Processing extension- and package-data in obscure ways
 - Detecting package-data automatically
 - Having a "clean" command
 - In general way too much executable code

Now, I understand where you're coming from: Keep your code dry, automate as much as possible, do the least amount of manual work. In general, these are good tipps however for `setup.py` I've found them to be counterproductive.

Be honest, who of you understood `setup.py` in its entirety? :-)

This PR does away with all of that: 

 - It assumes setuptools. Not your problem if people are running stuff from the stoneage
 - It lets PIP figure out the rest. Again, not your problem.
 - Let Cython cythonize the code
 - Declare every single bit of shipped data manually. Makes it very readable and explicit.
 - No cleaning. `find -name ... -delete` or `git clean -fdX` do the job just fine.
 - Almost no executable code

The result is clean, understandable and 1/3 the size of the previous implementation.

Now, regarding `statsmodels/tsa/statespace/_statespace.pyx`.

That file was generated from `statsmodels/tsa/statespace/_statespace.pyx.in` using tempita. However Cython currently does not allow a tempita-preprocessing step. This is very unfortunate and was causing a lot of headache for `setup.py`. Thus for the time being, I chose to simply commit the result of tempita, too. Not ideal, but just a small side effect to having the comfort of a clean setup routine.

Also, compiling `_statespace.pyx` from `_statespace.pyx.in` is not really hard, so doing it manually during development in that file should be OK.

A neat solution would be cython detecting `.in` files and preprocessing them, automatically.

Note that this PR isn't meant as "take it or leave it" but maybe also as a starting point for discussion.

### CODE REVIEW

**Comment by jseabold** [comment]
We'll need to keep the version.py writing stuff and keep the resultant file out of git.

**Comment by jseabold** [comment]
I suspect we're going to continue not to let pip handle dependencies, given it's default recursive upgrade behavior. Cf. long discussions about this.

**Comment by nils-werner** [comment]
It's the first time I've heard about PIP upgrade problems. Can you point me to the discussion you've had?

**Comment by josef-pkt** [comment]
This also makes cython a install dependency, for install based on an sdist, doesn't it?

Looks partially like a simplification by throwing away all the extras that we built into setup.py.

**Comment by josef-pkt** [comment]
What you call "antipatterns" in the descriptions, are mostly our "features".

Overall, we follow largely the scipy and pandas patterns for setup.py, with some of our own wrinkles.


-  Differentiating between distutils/setuptools (including versions of them)
   inherited from a long time ago, maybe irrelevant, 
-  Managing dependencies manually (instead of letting PIP do the heavy lifting)
   #1897 is maybe the main issue for the discussion but there are several others, "I hate pip" or requires in pip
-  Doing some weird git tag management
   seems a requirement, we (scipy, statsmodels) use it, we get an informative version number for the dev version
-  Manually cythonizing code (instead of letting Cython do the job)
   has the advantage that we don't need to have a sufficiently recent version of cython as install dependency from an `sdist`, seems to work well
-  Processing extension- and package-data in obscure ways
-  Detecting package-data automatically
   Has the advantage that we don't have to add new files to setup.py. But it's not always easy to figure out what's included and what's not.
-  Having a "clean" command
   no comment, I don't use it.
-  In general way too much executable code
   "too much" or "just what's needed"?

**Comment by bashtage** [comment]
IMO `setup.py` should should not allow pip to try and download/build NumPy and SciPy, in particular. These are hard to build on many platforms and will just lead to difficult to interpret errors if someone tries to install statsmodels.  It is far better to tell the user that these requirements are not met and let them make an explicit choice to install or upgrade.

**Comment by bashtage** [comment]
> Differentiating between distutils/setuptools (including versions of them) inherited from a long time ago, maybe irrelevant, 

Setuptools has moved on a lot, so probably a good idea to standardize 

> Having a "clean" command no comment, I don't use it.

Useful if working with different Cythons and Python versions using `python setup.py develop`.

File: statsmodels/version.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+short_version = '0.7.0'
```
Comment (bashtage) [inline statsmodels/version.py]: Version string in multiple places seems like a bad idea.

**Comment by bashtage** [comment]
Some additional reasons to drop distutils from setup

```
DEPRECATION: Uninstalling a distutils installed project (statsmodels) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
```

File: statsmodels/version.py (hunk 1 above)
Comment (nils-werner) [inline statsmodels/version.py]: Yup, but there are two alternatives:

 - Read `version.py` in `setup.py`.

  Requires readability of package data, something I disagree conceptually: you should not be required to touch your stuff during installation. Weird paths, weird environments, it might fail in weird situations you have no control over.
 - Write `version.py` in `setup.py`.

  Same as above but even worse: Each installation procedure changes the stuff it is about to install. So, potentially (again, in some weird edge cases), if something during install fails, you might end up with an installed package, but without `version.py`. No unit test can test for this.
- Keep them duplicate.

  Content and installation is separate. No executable code. Any redesign of the installation procedure (`setup.ini` instead of `setup.py` or whatever the future might bring) is a matter of an easy translation. Unit tests can test if `version.py` and `setup.py` agree about the version number.

**Comment by nils-werner** [comment]
> IMO setup.py should should not allow pip to try and download/build NumPy and SciPy, in particular. 

Yes, it is a pain but this should not be your concern. After all, you're just a package, not a package manager or a Linux distribution. Your docs might mention `pip install --no-deps statsmodels` instead.

**Comment by saketkc** [comment]
I agree with @bashtage, 
This has also been extensively discussed at few other places(one of the culprits being me, myself):
https://github.com/scikit-learn/scikit-learn/pull/4332#issuecomment-77403722

**Comment by nils-werner** [comment]
> Managing dependencies manually (instead of letting PIP do the heavy lifting) #1897 is maybe the main issue for the discussion but there are several others, "I hate pip" or requires in pip

Yeah, that's very unfortunate and I don't see a solution. But more fragmentation (each project using diffent hacks in their `setup.py`) only makes it harder for the PIP folks to find a solution that fits all.

> Detecting package-data automatically Has the advantage that we don't have to add new files to setup.py. But it's not always easy to figure out what's included and what's not.

I disagree. The current solution makes it super hard to add new files to setup.py because you need to dig in and understand the detection algorithm. A bunch of lines makes it immediately obvious where to add new stuff.

> In general way too much executable code "too much" or "just what's needed"?

I may be a hardliner here, but "any executable code" is too much in my book :-)

**Comment by nils-werner** [comment]
> I suspect we're going to continue not to let pip handle dependencies, given it's default recursive upgrade behavior. Cf. long discussions about this.

I've read something about `pip install --upgrade --no-deps package` somewhere but you've probably heard of it/discussed that already.

**Comment by bashtage** [comment]
> Yes, it is a pain but this should not be your concern. After all, you're just a package, not a package manager or a Linux distribution. Your docs might mention pip install --no-deps statsmodels instead.

I disagree.  User experience, especially for less-well-informed users, should be an important consideration.  A nice friendly error message that states why it cannot be installed is vastly better than a horrible nested pip crash because some complex upstream package can't be built.

I think it you look around scientific python `setup.py`s, you'll see quite a but of code execution as part of the process - SM is not unique here.

**Comment by nils-werner** [comment]
Definitely, user experience is important.

But imagine all the effort that goes into each individual scipy-related `setup.py` went into making PIP give a more friendly message instead...

**Comment by josef-pkt** [comment]
@nils-werner  Our effort has for years gone into our setup.py, and before that it was developed for numpy and scipy.

pip was developed for installing python packages into a (empty) virtualenv.
It didn't know how to install binaries which is crucial for Windows users (like I) and it just upgraded a core package like numpy **by default**, which in my opinion is close to upgrading python itself (not completely correct because numpy developers have been careful with binary compatibility in recent years).

Once pip is catching up enough, we will put more effort into this. For now I still largely ignore pip for my own use cases, except when using tox.
We (actually scipy developers) spend some effort accommodating pip and installation into an empty environment in our setup.py (adding more code).

**Comment by nils-werner** [comment]
Have you tried setting

    [tox]
    skipsdist = true

or maybe

    install_command=true {opts} {packages}

in your `tox.ini`? This may disable tox from fiddling around with your virtualenv.

**Comment by andreabedini** [comment]
commeting to keep myself in the loop. Please note that pip is the official python package manager and AFAIK it does inject setuptools when running setup.py, so we are all using setuptools anyway. 

> But imagine all the effort that goes into each individual scipy-related setup.py went into making PIP give a more friendly message instead...

I can't agree with this more...

**Comment by josef-pkt** [comment]
closing, I don't think we are going to do any wholesale simplification of our setup.py
